### PR TITLE
Feature / Achievements, holes, portals, finish-args, search, changelog

### LINKED ISSUE #3: Rank by number of "holes" (permissions requested) instead of stars
Having a rank of apps based on how many permissions each one require, would be much more interesting.

### PR DESCRIPTION
**Pull request by Arxcis**
Fixes #3

### Functional review help
1. Run python3 server
```
make server
```
2. Open browser at `0.0.0.0:8000/`
3. Click around

### Demo - Sort by holes

image

### CODE REVIEW

**Comment by Arxcis** [comment]
@tchx84 Maybe you want to review this PR? I would be grateful :pray:

**Comment by tchx84** [comment]
Ill wait for this PR to converge, since it's still changing rapidly, but I have one suggestion, don't mix data and code. If you generate data, just drop it in a json file and load it, but don't mix it with the code.

**Comment by Arxcis** [comment]
@tchx84 Which data are you specifically talking about. Do you mean the `db/flathub/*.js` files?

I wanted to dump them as `.json` files initially, but the browser ES modules does not accept `.json` as mimetype. At least when I tried in latest Firefox:

**Does not work**
```js
import count from "./data/flathub/count.json"
```
**Works**
```js
import count from "./data/flathub/count.js"
```
I tried to separate the data from the code, by putting in a separate `data`-folder.

**Update**
- Renamed the `db/`-folder -> `data/` in 6c802fc to make it clearer that it is data-only.
- Added `## Folder structure`-section to the README.md in e150fea

**Comment by tchx84** [comment]
> @tchx84 Which data are you specifically talking about. Do you mean the `db/flathub/*.js` files?

Yes, the proper way of doing this is: https://stackoverflow.com/questions/16991341/json-parse-file-path

**Comment by Arxcis** [comment]
> Yes, the proper way of doing this is: https://stackoverflow.com/questions/16991341/json-parse-file-path

Ah, of course! Thanks :+1: So in modern javascript that would look like:
```js
const res = await fetch("./data/flathub/count.json")
const count = await res.json()
```

File: index.html [hunk 1]
```diff
@@ -113,241 +113,247 @@ <h1 id="title" title="Flatstat | Graph">flatstat</h1>
...
-          fill: false,
-        },
-      ],
-    },
-    options: {
-      ...options,
-    },
-  });
+  async function renderCharts() {
+    const count = await useJSON("./data/flathub/count.json")
```
Comment (Arxcis) [inline index.html:74]: **Update:** Loading data as `.json` instead of `.js` here...

File: lists/index.html [hunk 2]
```diff
@@ -34,36 +34,20 @@ <h1 id="title" title="Flatstat | List">flatstat</h1>
...
+  window.addEventListener("hashchange", renderList)
+
 
-    const graphLink = document.getElementById(window.location.hash)
-    graphLink?.classList.add("current")
-    renderList();
-  })
+  // Map, filter, sort, slice, join list
+  async function renderList() {
+    const metafiles = await useJSON("../data/flathub/metafiles.json")
```
Comment (Arxcis) [inline lists/index.html]: ....and here

**Comment by Arxcis** [comment]
## Demo: Search and finish-args

**Comment by tchx84** [comment]
Looking great so far :+1:

**Comment by Arxcis** [comment]
@tchx84 Thanks! It was a good idea of you to wait with the review :sweat_smile: Things are changing a lot now in the early phase. Looking forward when things will stabilize :pray: I think I am getting close.

**Comment by Arxcis** [comment]
## Demo changelog

**Comment by Arxcis** [comment]
Ok, now I have to stop adding more features. This PR is getting too big.. :sweat_smile:

**Comment by Arxcis** [comment]
**Get a list of the new apps this month**

**Pull request merged**